code/controllers/subsystem/ambience.dm

```diff
@@ -7,21 +7,43 @@ SUBSYSTEM_DEF(ambience)
 	wait = 1 SECONDS
 	///Assoc list of listening client - next ambience time
 	var/list/ambience_listening_clients = list()
+	///Cache for sanic speed :D
+	var/list/currentrun = list()
 
 /datum/controller/subsystem/ambience/fire(resumed)
-	for(var/client/client_iterator as anything in ambience_listening_clients)
+	if(!resumed)
+		currentrun = ambience_listening_clients.Copy()
+	var/list/cached_clients = currentrun
 
-		if(isnull(client_iterator) || isnewplayer(client_iterator.mob))
-			ambience_listening_clients -= client_iterator
-			continue
+	while(cached_clients.len)
+		var/client/client_iterator = cached_clients[cached_clients.len]
+		cached_clients.len--
+		process_ambience_client(client_iterator)
 
-		if(ambience_listening_clients[client_iterator] > world.time)
-			continue //Not ready for the next sound
+		if(MC_TICK_CHECK)
+			return
 
-		var/area/current_area = get_area(client_iterator.mob)
+/datum/controller/subsystem/ambience/proc/process_ambience_client(client/to_process)
+	if(isnull(to_process) || isnewplayer(to_process.mob))
+		ambience_listening_clients -= to_process
+		return
 
-		var/sound = pick(current_area.ambientsounds)
+	if(ambience_listening_clients[to_process] > world.time)
+		return //Not ready for the next sound
 
-		SEND_SOUND(client_iterator.mob, sound(sound, repeat = 0, wait = 0, volume = 25, channel = CHANNEL_AMBIENCE))
+	var/area/current_area = get_area(to_process.mob)
 
-		ambience_listening_clients[client_iterator] = world.time + rand(current_area.min_ambience_cooldown, current_area.max_ambience_cooldown)
+	if(!current_area) //Something's gone horribly wrong
+		stack_trace("A [to_process.ckey]'s mob [to_process.mob] has somehowe ended up in nullspace. WTF did you do")
```

Comment:
```suggestion
		stack_trace("[to_process.ckey]'s mob [to_process.mob] has somehow ended up in nullspace. WTF did you do")
```

Though you might want `key_name` here instead.

```suggestion
		stack_trace("[key_name(to_process)] has somehow ended up in nullspace. WTF did you do")
```

Reply:
Yeah that'll do